File: Breadth First Search/List.c
//Patrick Stumps
//pstumps@ucsc.edu
//CMPS 101
//PA 2
//List ADT


// Libraries
#include <stdlib.h>
#include <stdio.h>

#include "List.h"


// List and Node structs

typedef struct NodeObj {
    int data;
    struct NodeObj* next;
    struct NodeObj* prev;
} NodeObj;

typedef NodeObj* Node;

typedef struct ListObj{
    Node head;
    Node tail;
    Node curr;
    int length;
    int index;
} ListObj;

//Node constructor and FreeNode
Node newNode(int data){
    Node n = malloc(sizeof(NodeObj));
    n->data = data;
    n->next = NULL;
    n->prev = NULL;
    return(n);
}

void freeNode(Node* pN){
    if(*pN != NULL && pN != NULL){
        free(*pN);
        *pN = NULL;
    }
}

//List constructor and FreeList

List newList(void){
    List L = malloc(sizeof(ListObj));
    L->head = NULL;
    L->tail = NULL;
    L->curr = NULL;
    L->index = -1;
    L->length = 0;
    return (L);
}

void freeList(List* pL){
    if(pL == NULL || *pL == NULL){
        return;
    }
    while(length(*pL) != 0){
        deleteFront(*pL);
    }
    free(*pL);
    *pL = NULL;
    /*if(*pL != NULL && pL != NULL){
        Node t = (*pL)->head;
        while(t != NULL){
            Node n = t;
            t = t->next;
            free(n);
        }
        free(*pL);
        *pL = NULL;
    }*/
}


//Access functions

int length(List L){
    if( L == NULL ){
        printf("Cannot call length on empty or null List.");
        exit(1);
    }
    return(L->length);
}

int index(List L){
    if(L == NULL){
        printf("Cannot call index on NULL list.");
        exit(1);
    }

    return (L->index);
}

int front(List L){
    if(L == NULL){
        printf("List does not exist.");
        exit(1);
    }
    if(length(L) == 0){
        printf("Cannot call front on empty list.");
        exit(1);
    }
    return(L->head->data);
}

int back(List L){
    if(L == NULL){
        printf("List does not exist.");
        exit(1);
    }
    if(length(L) == 0){
        printf("Cannot call back on empty list.");
        exit(1);
    }
    return(L->tail->data);
}

int get(List L){
    if(L == NULL){
        printf("Cannot call get on NULL list.");
        exit(1);
    }
    if(length(L) == 0){
        printf("Cannot call get on empty list.");
        exit(1);
    }
    if(L->curr == NULL){
        printf("Cannot call get on NULL cursor.");
        exit(1);
    }
    if(L->curr < 0){
        printf("Cannot call get on NULL cursor (cursor index < 0");
        exit(1);
    }
    return(L->curr->data);
}

int equals(List A, List B){
    Node cmp = A->head;
    Node th =  B->head;
    while(cmp->next != NULL){
        if(cmp->data != th->data){
            return 0;
        }
        th = th->next;
        cmp = cmp->next;
    }
    return 1;
}

//Manipulation procedures

void clear(List L){
    if(L == NULL){
        printf("Cannot clear NULL or empty list.");
        exit(1);
    }
    //if(length(L) == 0){
    //    printf("Cannot clear empty list. (length = 0)");
    //    exit(1);
    //}
    while(L->head != NULL){
        deleteFront(L);
    }
    L->head = NULL;
    L->tail = NULL;
    L->curr = NULL;
    L->length = 0;
    L->index = -1;
}

void moveFront(List L){
    if(L == NULL){
        printf("Cannot moveFront NULL list.");
        exit(1);
    }
    if(L->length > 0){
        L->curr = L->head;
        L->index = 0;
    }
}

void moveBack(List L){
    if(L == NULL){
        printf("Cannot moveBack NULL list.");
        exit(1);
    }
    if(L->length > 0){
        L->curr = L->tail;
        L->index = L->length -1;
    }
}

void movePrev(List L){
    if(L->curr == NULL){
        L->curr = NULL;
        L->index = -1;
    }
    else if (L->curr != NULL && L->curr == L->head){
        L->curr = NULL;
        L->index = -1;
    }
    else if(L->curr != NULL && L->curr != L->head){
        L->curr = L->curr->prev;
        L->index--;
    }
    /*if(index(L) == -1){
        return;
    }
    if(length(L) == 0){
        L->curr = NULL;
    }
    if(index(L) >= 0 && index(L) >= length(L) -1){
        L->curr = L->curr->prev;
    }*/

}

void moveNext(List L){
    if(L->curr == NULL){
        L->curr == NULL;
        L->index = -1;
    }
	else if (L->curr != NULL && L->curr == L->tail){
    	L->curr = NULL;
        L->index = -1;
	}
    else if(L->curr != NULL && L->curr != L->tail){
        L->curr = L->curr->next;
        L->index++;
    }
}

void prepend(List L, int data){
    if(L == NULL){
        printf("Cannot prepend NULL list");
        exit(1);
    }
    Node N;
    if (length(L) == 0){
        L->head = newNode(data);
        L->tail = L->head;
        L->curr = L->head;
    }
    else{
        N = L->head;
        N->prev = newNode(data);
        L->head = N->prev;
        L->head->next = N;
        L->index++;
    }
    L->length++;
}

void append(List L, int data){
    if(L == NULL){
        printf("Cannot append NULL list");
        exit(1);
    }
    Node N;
    if(length(L) == 0){
        L->tail = newNode(data);
        L->head = L->tail;
        L->curr = L->tail;
    }
    else{
        N = L->tail;
        N->next = newNode(data);
        L->tail = N->next;
        L->tail->prev = N;
    }
    L->length++;
}

void insertBefore(List L, int data){
    // BUG: Calling when L->curr == NULL.
    if(L->curr == NULL){
        printf("Cannot call insertBefore on NULL cursor.");
        exit(1);
    }
    Node N = newNode(data);
    if(L->curr == L->head){
        /*L->head->prev = N;
        N->next = L->head;
        L->head = N;*/
        prepend(L, data);
        return;
    }
    else{
        N->prev = L->curr->prev;
        L->curr->prev->next = N;
        N->next = L->curr;
        L->curr->prev = N;
    }
    L->length++;
    /*if(index(L) == 0){
        prepend(L, data);
        return;
    }
    N->prev = L->curr->prev;
    N->next = L->curr;
    L->curr->prev->next = N;
    L->curr->prev = N;
    L->length++;*/
}

void insertAfter(List L, int data){
    // BUG: Calling when L->curr == NULL.
    if(L->curr == NULL){
        printf("Cannot call insertAfter on NULL cursor.");
        exit(1);
    }
    Node N = newNode(data);
    if(L->curr == L->tail){
        L->tail->next = N;
        N->prev = L->curr;
        L->tail = N;
    }
    else{
        N->prev = L->curr;
        N->next = L->curr->next;
        L->curr->next->prev = N;
        L->curr->next = N;
    }
    L->length++;
    /*if(index(L) == length(L)){
        append(L, data);
        return;
    }
    N->prev = L->curr;
    N->next = L->curr->next;
    L->curr->next->prev = N;
    L->curr->next = N;
    L->length++;*/
}

void deleteFront(List L){
    if(L == NULL){
        printf("Cannot call deleteFront on NULL list.");
        exit(1);
    }
    if(length(L) < 0){
        printf("Cannot call deleteFront on empty list.");
        exit(1);
    }
    Node N = L->head;
    if(L->length > 1){
        L->head = L->head->next;
        freeNode(&N);
        L->head->prev = NULL;
        L->index--;
    }
    else{
        L->head = L->tail = NULL;
        L->curr = NULL;
        L->index--;
        freeNode(&N);
    }
    L->length--;
}

void deleteBack(List L){
    if(L == NULL){
        printf("Cannot call deleteBack on NULL list.");
        exit(1);
    }
    if(L->length == 0 || L->length < 0){
        printf("Cannot call deleteBack on empty list.");
        exit(1);
    }
    Node N = L->tail;

    if (length(L) > 1) {
        L->tail = L->tail->prev;
        L->tail->next = NULL;
        freeNode(&N);
    } 
    if (length(L) == 1) {
        L->head = NULL;
        L->tail = NULL;
        L->length = 0;
        L->index = -1;
        freeNode(&N);
    }
    if(L->curr == L->tail){
        L->index = -1;
        L->curr = L->curr->prev;
        freeNode(&N);
        L->index--;
    }
}

void delete(List L){
    if(L == NULL){
        printf("Cannot call delete on NULL list.");
        exit(1);
    }
    if(L->length == 0){
        printf("Cannot call delete on empty list.");
        exit(1);
    }
    if(L->length < 0){
        printf("Cannot call delete on negative length list.");
        exit(1);
    }
    if(L->index < 0){
        printf("Cannot call delete on null cursor.");
        exit(1);
    }
    Node N = L->curr;

    if(N == L->tail){
        deleteBack(L);
    }
    else if(N == L->head){
        deleteFront(L);
    }
    else{
        L->curr->prev->next = L->curr->next;
        L->curr->next->prev = L->curr->prev;
        //L->curr->next = NULL;
        //L->curr->prev = NULL;
        freeNode(&N);
        L->curr = NULL;    
        L->length--;
        L->index = -1;
    }
}

//Other Functions

void printList(FILE* out, List L){
   if(L == NULL){
       printf("Cannot call printList on NULL list");
       exit(1);
   }
   for(Node N = L->head; N != NULL; N = N->next){
       fprintf(out, "%d ", N->data);
   }
}

List copyList(List L){
    List l = newList();
    for (Node N = L->head; N != NULL; N = N->next){
        append(l, N->data);
    }
    return (l);
}

File: Lex-Java/List.java
//Patrick Stumps
//pstumps@ucsc.edu
//CMPS 101
//PA 1
//List ADT
public class List {
	
	public class Node {
		int data;
		int index;
		Node next;
		Node prev;
		
		Node(int d){
			data = d;
			next = null;
			prev = null;
		}
		
		public String toString() {
			return String.valueOf(data);
		}
	}
	
	private Node head;
	private Node tail;
	private Node curr;
	private int length;
	private int index;
	
	//Constructor
	
	public List() {
		head = null;
		tail = null;
		curr = null;
		length = 0;	
	}
	
	//Access Functions
	
	int length() {
		return length;
	}
	
	//returns front element
	int front() {
		if (length == 0) throw new RuntimeException
			("Front does not exist.");
		return head.data;
	}
	
	//return back element
	int back() {
		if (length == 0) throw new RuntimeException
			("Tail does not exist.");
		return tail.data;
	}
	
	//returns cursor element
	int get() {
		if(length() > 0 && index() >= 0){
			return curr.data;
		}
		return -1;
	}
	
	//returns cursor of index element
	int index() {
		if (curr == null) {
			return -1;
		}
		Node t = head;
		int index = 0;
		while( t!= null){
			if(t.data == curr.data){
				return index;
			}
			index++;
			t=t.next;
		}
		return -1;
	}
	
	boolean equals (List L) {
		Node cmp = L.head;
		Node th = head;
		while(cmp.next != null) {
			if(cmp.data != th.data) {
				return false;
			}
			cmp = cmp.next;
			th = th.next;
			
		}
		return true;
	}
	
	// Manipulation procedures
	
	void clear() {
		head = null;
		length = 0;
		index = -1;
	}
	//places cursor under front element
	void moveFront() {
		if(length > 0){
			curr = head;
		}
	}
	//places cursor under back element
	void moveBack() {
		if(length > 0){
			curr = tail;
		}
	}
	//moves cursor one step toward front of list
	void movePrev() {
		if(index() == -1){
			return;
		}
		if(index() == 0){
			curr = null;
		}
		if(index() >= 0 && index() <= length -1){
			curr = curr.prev;
		}
	}
	//moves cursor one step toward back
	void moveNext() {
		if(curr != null && index() != length -1){
			curr = curr.next;
		}
		else if (curr != null && index() == length-1){
			curr = null;
		}
	}
	//insert new element into front of list
	void prepend(int data) {
		Node N;
		if (length == 0) {
			head = new Node(data);
			tail = head;
			curr = head;
		}
		else {
			N = head;
			N.prev = new Node(data);
			head = N.prev;
			head.next = N;
		}
		//for (Node curr = tail.prev; curr != null; curr=curr.prev) {
		//	++curr.index;
		//}
		length++;
	}
	//insert new element after back of list
	void append(int data) {
		Node N;
		if (length == 0) {
			tail = new Node(data);
			head = tail;
			curr = tail;
		}
		else {
			N = tail;
			N.next = new Node(data);
			tail = N.next;
			tail.prev = N;
		}
		//for (Node curr = head.next;curr != null; curr = curr.next) {
		//	++curr.index;
		//}
		length++;
	}
	//insert element before cursor
	void insertBefore(int data) {
		Node N = new Node(data);
		if (index() == 0) {
			prepend(data);
			return;
		}
		Node cprev = curr.prev;
		N.prev = cprev;
		N.next = curr;
		cprev.next = N;
		curr.prev = N;
		length++;
	}
	//insert element after cursor
	void insertAfter(int data) {
		Node N = new Node(data);
		if (length == 0) throw new RuntimeException
			("Node does not exist");
		
		N.prev = curr;
		N.next = curr.next;
		curr.next = N;
		length++;
	}
	
	void deleteFront() {
		if (length == 0) throw new RuntimeException
			("Head node does not exist.");
		head = head.next;
		head.next = head;
		length--;
	}
	
	void deleteBack() {
		if (length == 0) throw new RuntimeException
			("Back node does not exist.");
		tail = tail.prev;
		tail.prev = tail;
		length--;
	}
	
	void delete() {
		if (length == 0) throw new RuntimeException
			("Node does not exist.");
		curr = null;
		length--;
	}
	
	// Other Methods
	
	public String toString() {
		String l = "";
		for (Node curr = head; curr != null; curr = curr.next) {
			l = l += curr.toString() + " ";
		}
		return l;
	}
	
	List copy() {
		List l = new List();
		for (Node curr = head; curr != null; curr = curr.next) {
			l.append(curr.data);
		}
		return l;
	}
	
}